### DISCUSSION TITLE
How to check if a point is generator in an elliptic curve?

### DISCUSSION THREAD

**Discussion by erhant** [erhant/moonmath#7] [Ideas]
In cyclic groups, I could use co-factor clearing to check if a point is a generator:

```py
def find_generators(p):
  Fp = GF(p)

  order = p - 1
  factorization = factor(order)
  print("Factors of", order, "are", factorization)

  gens = []
  for e in range(2, p):
    e = Fp(e)

    is_gen = True
    for f in map(lambda f: f[0], factorization):
      cf = order // f # cofactor
      if e ** cf == 1:
        # cofactor clearing results in 1, this is NOT a generator
        is_gen = False
        break

    if is_gen:
      gens.append(e)

  print(gens)
  assert(len(gens) == euler_phi(order))
```

Is there something analogous to this one for Elliptic Curves?

**Comment by skaunov** [comment]
I could be wrong, but I feel like the same should work for EC too, since we just check if the _group operation_ applied _subgroup order_ times to `e` yields identity element.

So I'd just take EC order for `order` and scalar multiplication instead of _power_.

**Comment by skaunov** [comment]
So...
In <https://github.com/LeastAuthority/moonmath-manual/blob/158934a9e4452ae127dae0814fb43be2816cd9c1/chapters/algebra-moonmath.tex#L211> _cofactor clearing_ is described in relation to _finite cyclic groups_ (and not restricted to multiplicative groups of integers modulo a prime). (And details are outlined in the [definition](https://github.com/LeastAuthority/moonmath-manual/blob/158934a9e4452ae127dae0814fb43be2816cd9c1/chapters/algebra-moonmath.tex#L202C2-L202C2) in the beginning of the subsection.)
Next, it's stated that ["It can be shown that this group is finite and cyclic whenever the underlying field $\F$ is finite."](https://github.com/LeastAuthority/moonmath-manual/blob/158934a9e4452ae127dae0814fb43be2816cd9c1/chapters/elliptic-curves-moonmath.tex#L414), consequently _cofactor clearing_ should be applicable to EC as well. Cheers!
